consult.el

```diff
@@ -3259,14 +3278,41 @@ IGNORED-FACES are ignored when determining the match position."
   "Lookup position of match.
 SELECTED is the currently selected candidate.
 CANDIDATES is the list of candidates.
-INPUT is the input string entered by the user."
-  (consult--line-point-placement selected candidates
-                                 (and (not (string-blank-p input))
-                                      (car (consult--completion-filter
-                                            input
-                                            (list (substring-no-properties selected))
-                                            'consult-location 'highlight)))
-                                 'completions-first-difference))
+INPUT is the input string entered by the user.
+Return value is a list of the form
+(<line pos> (<match start> . <match end>)* <line pos>...)
+where the match offsets are relative to the previous line pos,
+and the selected candidate comes first.
+"
+  (when (not consult-line-preview-all-matches)
+    (setq candidates (list (car (member selected candidates)))))
+
+  (let ((candidates (consult--completion-filter
+                     input
+                     (mapcar (lambda (cand)
+                               (propertize
+                                (substring-no-properties cand)
+                                'consult-location (get-text-property 0 'consult-location cand)))
+                               candidates)
+                     'consult-location 'highlight)))
```

Comment:
Thanks for this!

A quick comment about performance - I am unlikely to accept this patch in this form since an entire list of completion candidates and all the candidate strings (!) are allocated here on every key press for preview. This will be too costly for the Emacs GC.

Also note that I am not happy to add this special casing only for consult-line/consult-outline. There are also commands like consult-grep, consult-line-multi and consult-info which would need a similar treatment.

Reply:
Yes, this function is implemented in an awkward spot. Having to strip properties before passing candidates through the matcher is problematic. I wanted to first cheaply filter to visible candidates, but that can't happen here for multiple reasons.

The whole approach of passing the lines through the matcher (again) to find the highlight ranges seems very roundabout. Even the current implementation is `O(total candidates)` which conceptually doesn't seem necessary.

Would something like this be possible?
1. Somehow hook the minibuffer completion matching to record the match ranges as an explicit string property.
1. Pass this function a list of already filtered instead of all candidates (optional - could use above match range metadata to efficiently filter).
1. Propagate match ranges without having to do any real work.

Reply:
> Even the current implementation is O(total candidates) which conceptually doesn't seem necessary.

You mean the candidate lookup? While that's true, it is not that relevant considering the constants. Allocating all the strings is a different story.

> The whole approach of passing the lines through the matcher (again) to find the highlight ranges seems very roundabout.

It is the best we came up with while sticking to the exposed APIs.

> Would something like this be possible?

Yes, with some significant hacking, since you can do almost everything with advices.

I think I didn't emphasize this enough - I am happy with the status quo. It nicely highlights the current line I am looking at. I don't miss the additional highlights. Note that consult-line also has other issues like #281, which are difficult to fix if one doesn't want to hack everything deeply. If you are absolutely relying on preserving the Swiper behavior, why not continue using it?

If it is possible to figure out a generalization which fits nicely into the given framework I am happy to consider it of course. There are two factors at play here - one side is the UI (which you want to improve) and one side is the implementation, which I want to keep simple and which should preserve API boundaries. The goal is to avoid entangling the UI and Consult.

Reply:
> Somehow hook the minibuffer completion matching to record the match ranges as an explicit string property.

Probably can't do this in general, but looks like I can do it easily enough w/ orderless.

Reply:
>  If you are absolutely relying on preserving the Swiper behavior, why not continue using it?

It's a good question. I was never fully satisfied w/ Swiper, even after a bunch of hacking. In general I believe with a few tweaks a line based approach like swiper or consult-line can cover 95% of my use of isearch with an overall improved UX.